include/sys/scheduler.h

```diff
@@ -0,0 +1,171 @@
+				void *swap_data)
+{
+	return k_sched_callback_wake(wait_q, swap_retval, swap_data,
+				     NULL, NULL, NULL);
+}
+
+/**
+ * Wake up all threads pending on the provided wait queue
+ *
+ * Convenience function to invoke k_sched_wake() on all threads in the queue
+ * until there are no more to wake up.
+ *
+ * @param wait_q Wait queue to wake up the highest prio thread
+ * @param swap_retval Swap return value for woken thread
+ * @param swap_data Data return value to supplement swap_retval. May be NULL.
+ * @retval true If any threads were woken up
+ * @retval false If the wait_q was empty
+ */
+static inline bool k_sched_wake_all(_wait_q_t *wait_q, int swap_retval,
+				    void *swap_data)
+{
+	bool woken = false;
+
+	while (k_sched_wake(wait_q, swap_retval, swap_data)) {
+		woken = true;
+	}
+
+	/* True if we woke at least one thread up */
+	return woken;
+}
+
+/**
+ * Atomically put the current thread to sleep on a wait queue, with timeout
+ *
+ * The thread will be added to the provided waitqueue. The lock, which should
+ * be held by the caller with the provided key, will be released once this is
+ * completely done and we have swapped out.
+ *
+ * The return value and data pointer is set by whoever woke us up via
+ * k_sched_wake.
+ *
+ * @param lock Address of spinlock to release when we swap out
+ * @param key Key to the provided spinlock when it was locked
+ * @param wait_q Wait queue to go to sleep on
+ * @param timeout Waiting period to be woken up, or K_FOREVER to wait
+ *                indefinitely.
+ * @param data Storage location for data pointer set when thread was woken up.
+ *             May be NULL if not used.
+ * @retval Return value set by whatever woke us up, or -EAGAIN if the timeout
+ *         expired without being woken up.
+ */
+int k_sched_wait(struct k_spinlock *lock, k_spinlock_key_t key,
+		 _wait_q_t *wait_q, k_timeout_t timeout, void **data);
+
+/**
+ * Atomically invoke a scheduling decision
+ *
+ * Waking up a thread with APIs like k_sched_wake() un-pends and readies
+ * the woken up thread, but does not implicitly invoke the scheduler to
+ * immediately schedule them. This allows other tasks to be performed in
+ * between.
+ *
+ * When called, the kernel may context switch to some other thread,
+ * releasing the lock once we have switched out. The lock will never
+ * be held when this returns.
+ *
+ * If the caller is a co-operative thread, or the next thread to run on this
+ * CPU is otherwise still the current thread, this just unlocks the spinlock
+ * and returns.
+ *
+ * Most users will use the same lock here that synchronizes k_sched_wait()
+ * and k_sched_wake() calls.
+ *
+ * It is safe (albeit very slightly wasteful) to call this even if no thread
+ * was woken up.
+ *
+ * @param lock Address of spinlock to release when we swap out
+ * @param key Key to the provided spinlock when it was locked
+ */
+void k_sched_invoke(struct k_spinlock *lock, k_spinlock_key_t key);
```

Comment:
This would be nice: I think it could be the thing that gets invoked at a [reschedule point](https://docs.zephyrproject.org/latest/reference/terminology.html#api-term-reschedule).

Reply:
this particular API is the reschedule point. it is currently a thin wrapper on z_reschedule().